Add discussion of main actor and @globalActor [SE-0316] (#366)
swiftlang/swift-book: TSPL.docc/LanguageGuide/Concurrency.md

A program that uses parallel and asynchronous code
carries out multiple operations at a time,
and it suspends operations that are waiting for an external system.
The rest of this chapter uses the term *concurrency*
to refer to this common combination of asynchronous and parallel code.

The additional scheduling flexibility from parallel or asynchronous code
also comes with a cost of increased complexity.
When you write concurrent code,
you don't know ahead of time what code will run at the same time,
and you might not always know the order that code will run.
A common problem in concurrent code happens
when multiple pieces of code try to access
some piece of shared mutable state —
this is known as a *data race*.
When you use the language-level support for concurrency,
Swift detects and prevents data races,
and most data races produce a compile-time error.
Some data races can't be detected until your code is running;
these data races terminate code execution.

> <!-- XXX OUTLINE

Reviewer: Is this a note about work to come, or does it reference completed work? If it's completed, probably best to delete the note.

Author: Removed— the new section it refers to was separated onto the 'data_triangle_150692546' branch, to land later.

The build script for TSPL considers any appearances of XXX in source as a publication blocker.  They all have to be removed before it will let me build and ship the book.

}
```

<!--
XXX TR: Do we need to mention this?

> How common is it for developers to need to insert manual yields?

Reviewer 2: Inserting yield() manually is definitely not the first thing people need to encounter as they read up about concurrency, so I'd agree that we can move it out from here. 

I'd stuff it all the way at the end of the page maybe in some "forcing suspensions" section which could be something about debugging performance and "overhangs" with swift concurrency perhaps.

[`Task.checkCancellation()`]: https://developer.apple.com/documentation/swift/task/3814826-checkcancellation
[`Task.isCancelled` type]: https://developer.apple.com/documentation/swift/task/iscancelled-swift.type.property

> <!-- XXX is the Optional<Data>.self still needed? -->

Reviewer 2: Yeah, the `return nil` otherwise fails to infer that the closure of the task is `Data?`-returning (verified just now).

If we'd reshape the code inside to something like...

```
return nil as Data?
```

that would get rid of the need for the `(of: Optional<Data>.self)` as well... Alternatively we could:

```swift
        for name in photoNames {
            let added = group.addTaskUnlessCancelled {
                Task.isCancelled ? nil : await downloadPhoto(named: name)
            }
            guard added else { break }
        }

```

sadly it does not work with an if expression:

```
// sadly does not compile

           let added = group.addTaskUnlessCancelled {
                if !Task.isCancelled {
                    await downloadPhoto(named: name)
                } else {
                    nil // Cannot convert value of type 'ExpressibleByNilLiteral' to specified type 'Void'
                }
            }
```

You have complete flexibility to manage unstructured tasks
in whatever way your program needs,
but you're also completely responsible for their correctness.

To create an unstructured task
that runs similarly to the surrounding code,
call the [`Task.init(priority:operation:)`][] initializer.
The new task defaults to running with
the same actor isolation, priority, and task-local state as the current task.
To create an unstructured task
that's more independent from the surrounding code,
known more specifically as a *detached task*,
call the [`Task.detached(priority:operation:)`][] static method.

> The new task defaults to running as a nonisolated function

Reviewer 3: In the SE-0461 world, saying it runs as a nonisolated function has no implications about where it runs. Do we want to say it runs as an `@concurrent` function, or some other description of running off of an actor by default?

  (Pull from my 2021-04-21 notes from Ben's talk rehearsal.)
-->

## The Main Actor

The previous sections discuss approaches for dividing up concurrent work.
In some cases,
after divided-up work finishes,
you also need to join execution back together
to do some synchronous work model.
For example,
when an app updates its user interface,
changes need to be performed one at a time,
with each update completing before the next one begins,
to prevent data races.
UI updates can also come from many places in your code,
so grouping all UI-related code on a single type
doesn't work very well.
<!-- XXX TR: Other examples of non-UI non-main-actor work? -->
To serialize concurrent work like this,
you can use the *main actor*,

> which is a shared instance of [`MainActor`][].

Reviewer 4: Hmm. I don't think I would center this around "joining execution back together". It looks like TSPL doesn't currently have a good introduction to isolation as a general concept. I think that's what you're trying to do here, but I worry that attaching that discussion to main actors is actually going to be more confusing than not. Let me make a counter-suggestion to show what I'm thinking.

```
## Isolation

The previous sections discuss approaches for splitting up concurrent work.
That work may involve making changes to shared data, such as an app's UI.
If different parts of your code could modify the same data at the same time,
that would create a *data race*, and your program would not behave correctly.

Swift doesn't let that happen. Whenever you read or modify a piece of
data, Swift ensures that no other code is modifying it concurrently.
This is called *data isolation*. There are three main ways that data
can be isolated:

First, immutable data is always isolated. No code can possibly be
modifying a `let` constant at the same time you're reading it
because no code can modify it at all.

Second, data that's only referenced by the current task is always
isolated. A local `var` is safe to read and write because no other
code has a reference to that memory. (If you capture the `var` in
a closure, Swift will make sure the closure isn't used concurrently.)

Finally, data that's protected by an actor is isolated if the current
code is isolated to that actor. An actor is an object that protects
access to mutable data by forcing code to take turns. If the current
function is isolated to an actor, it is safe to read and write data
that's protected by that actor because any other code that's
isolated to that same actor must wait its turn before it can run.

## The main actor

The most important actor in many programs is the main actor.
In an app, the main actor protects all of the data that's used to
show the UI. The main actor takes turns rendering the UI,
handling UI events, and running code that you write that
needs to query or update the UI.

<!-- explain how to use the `@MainActor` attribute here -->
<!-- explain how Swift restricts uses of isolated data, calls to
   @MainActor functions, etc. -->

<!-- wait for SE-0478 to give the official syntax for this -->
When you're writing UI code, you often want all of it to be
isolated to the main actor. To do this, you can write
`using @MainActor` at the top of a Swift file to make the
attribute apply by default to all the code in the file.

If there's a specific function or property that you want to
exclude from `using @MainActor`, you can use the
`nonisolated` modifier on that declaration to override the
default.

Modules can be configured to be built using `using @MainActor`
by default. This can be overridden on a per-file basis by
writing `using nonisolated` at the top of a file.

## Actor classes

<!-- put most of the existing section about `actor` here,
      emphasizing the parallels between this and global actors -->

## Other global actors

The main actor is actually just a global singleton instance of
the `MainActor` actor class. An actor class can normally have
multiple instances, each providing independent isolation.
This is why all the isolated data of the actor must be declared
as instance properties of the actor class. However, since
`MainActor` is singleton, the type alone is sufficient to identify
the actor. This allows main actor isolation to be declared with
just an attribute, which gives you much more flexibility to
organize your program as you like.

You can define your own singleton global actors using the
`@globalActor` attribute.

<!-- docs here -->
```

Anyway, let me know what you think.

Author: Thanks, John!  I agree — adding a section to define isolation here makes it easier to explain actors and the main actor, and fills in a conceptual gap in the book.

  (Pull from my 2021-04-21 notes from Ben's talk rehearsal.)
-->

## The Main Actor

The previous sections discuss approaches for dividing up concurrent work.
In some cases,
after divided-up work finishes,
you also need to join execution back together
to do some synchronous work model.
For example,
when an app updates its user interface,
changes need to be performed one at a time,
with each update completing before the next one begins,
to prevent data races.
UI updates can also come from many places in your code,

> so it wouldn't work very well to

Reviewer: ```suggestion
so grouping all UI-related code on a single type doesn't work very well.
```

> try grouping all UI-related code on a single type.

Reviewer: ```suggestion
```

<!-- XXX TR: Other examples of non-UI non-main-actor work? -->
To serialize concurrent work like this,
you can use the *main actor*,
which is a shared instance of [`MainActor`][].

[`MainActor`]: https://developer.apple.com/documentation/swift/mainactor

Before you start using concurrency in your code,
everything runs on the main actor.
As you identify long-running or resource-intensive code,

> you can move this work off the main actor,

Reviewer: ```suggestion
you can move this work off the main actor
```

in a way that's still safe and correct.

> Note:
> The main actor is closely related to the main thread,
> but they're not the same thing.
> The main actor has private mutable state,
> and the main thread serializes access to that state.
> When you run code on the main actor,
> Swift executes that code on the main thread.
> Because of this connection,
> you might see these two terms used interchangeably.
> Your code interacts with the main actor;

> > the main thread is an lower level implementation detail.

Reviewer: ```suggestion
> the main thread is a lower-level implementation detail.
```

There are several ways to run work on the main actor.
To ensure a function always runs on the main actor,

> you mark it with the `@MainActor` attribute:

Reviewer: ```suggestion
mark it with the `@MainActor` attribute:
```

```swift
@MainActor
func show(_: Data) {
    // ... UI code to display the photo ...
}
```

In the code above,
the `@MainActor` attribute on the `show(_:)` function
requires this function to run only on the main actor.
Within other code that's running on the main actor,
you can call `show(_:)` as a synchronous function.
However,
to call `show(_:)` from code that isn't running on the main actor,
you have to include `await` and call it as an asynchronous function
because switching to the main actor introduces a potential suspension point.
For example:

```swift
func downloadAndShowPhoto(named name: String) async {
    let photo = await downloadPhoto(named: name)
    await show(photo)
}
```

In the code above,
both the `downloadPhoto(named:)` and `show(_:)` functions

> could suspend when you call them.

Reviewer: ```suggestion
might suspend when you call them.
```

This code also shows a common pattern:
Perform long-running and CPU-intensive work in the background,
and then switch to the main actor to update the UI.
Because the `downloadAndShowPhoto(named:)` function isn't on the main actor,
the work in `downloadPhoto(named:)` also doesn't run on the main actor.
Only the work in `show(_:)` to update the UI runs on the main actor,
because that function is marked with the `@MainActor` attribute.
<!-- TODO
When updating for SE-0461,
this is a good place to note
that downloadPhoto(named:) runs
on whatever actor you were on when you called it.
-->

> To require that a closure must run on the main actor,

Reviewer: ```suggestion
To require that a closure runs on the main actor,
```
